apps/src/lib/client/utils.rs

```diff
@@ -44,161 +44,177 @@ pub const NET_OTHER_ACCOUNTS_DIR: &str = "other";
 const RELEASE_PREFIX: &str =
     "https://github.com/heliaxdev/anoma-network-config/releases/download";
 
+fn log_and_exit<E: std::error::Error>(error: E) {
+    eprintln!("ERROR: {}", error);
+    tracing::error!("{:?}", error);  // ?: is using {:?} the best way to pass the error to tracing?
+    cli::safe_exit(1);
+}
+
+// TODO: can these two exit_if_error helpers be combined?
+fn exit_if_error_ref<E: AsRef<dyn std::error::Error>>(result: Result<(), E>) {
+    if let Err(error) = result {
+        log_and_exit(error.as_ref());
+    }
+}
+
+fn exit_if_error<E: std::error::Error>(result: Result<(), E>) {
+    if let Err(error) = result {
+        log_and_exit(error);
+    }
+}
+
 /// Configure Anoma to join an existing network. The chain must be released in
 /// the <https://github.com/heliaxdev/anoma-network-config> repository.
 pub async fn join_network(
     global_args: args::Global,
     args::JoinNetwork { chain_id }: args::JoinNetwork,
 ) {
-    use tokio::fs;
+    exit_if_error_ref(async {
+        use tokio::fs;
 
-    let base_dir = &global_args.base_dir;
-    let wasm_dir = global_args.wasm_dir.as_ref().cloned().or_else(|| {
-        if let Ok(wasm_dir) = env::var(ENV_VAR_WASM_DIR) {
-            let wasm_dir: PathBuf = wasm_dir.into();
-            Some(wasm_dir)
-        } else {
-            None
-        }
-    });
-    if let Some(wasm_dir) = wasm_dir.as_ref() {
-        if wasm_dir.is_absolute() {
-            eprintln!(
-                "The arg `--wasm-dir` cannot be an absolute path. It is \
-                 nested inside the chain directory."
-            );
-            cli::safe_exit(1);
-        }
-    }
-    if let Err(err) = fs::canonicalize(base_dir).await {
-        if err.kind() == std::io::ErrorKind::NotFound {
-            // If the base-dir doesn't exist yet, create it
-            fs::create_dir_all(base_dir).await.unwrap();
-        }
-    } else {
-        // If the base-dir exists, check if it's already got this chain ID
-        if fs::canonicalize(base_dir.join(chain_id.as_str()))
-            .await
-            .is_ok()
-        {
-            eprintln!("The chain directory for {} already exists.", chain_id);
-            cli::safe_exit(1);
+        let base_dir = &global_args.base_dir;
+        let wasm_dir = global_args.wasm_dir.as_ref().cloned().or_else(|| {
+            if let Ok(wasm_dir) = env::var(ENV_VAR_WASM_DIR) {
+                let wasm_dir: PathBuf = wasm_dir.into();
+                Some(wasm_dir)
+            } else {
+                None
+            }
+        });
+        if let Some(wasm_dir) = wasm_dir.as_ref() {
+            if wasm_dir.is_absolute() {
+                eprintln!(
+                    "The arg `--wasm-dir` cannot be an absolute path. It is \
+                    nested inside the chain directory."
+                );
+                cli::safe_exit(1);
```

Comment:
this could be changed to returning something like `Err("The arg `--wasm-dir` cannot be an absolute path. It is nested inside the chain directory.")`